src/Features/Core/Portable/AddConstructorParametersFromMembers/AddConstructorParametersFromMembersCodeRefactoringProvider.cs

```diff
@@ -56,25 +57,55 @@ public async Task<ImmutableArray<CodeAction>> AddConstructorParametersFromMember
+            else
+            {
+                // Create sub menus for suggested actions, one for required parameters and one for optional parameters
+                var requiredParameterCodeActions = new ArrayBuilder<CodeAction>();
+                var optionalPrameterCodeActions = new ArrayBuilder<CodeAction>();
+                foreach (var constructor in state.ConstructorCandidates)
+                {
+                    if (CanHaveRequiredParameters(constructor._missingParameters))
+                    {
+                        requiredParameterCodeActions.Add(new AddConstructorParametersCodeAction(document, constructor, containingType, constructor._missingParameters, useSubMenuName: true));
+                    }
+                    optionalPrameterCodeActions.Add(GetOptionalContructorParametersCodeAction(document, constructor, containingType, useSubMenuName: true));
+                }
 
-            var missingParameters = state.MissingParameters.SelectAsArray(p => CodeGenerationSymbolFactory.CreateParameterSymbol(
-                attributes: default,
-                refKind: p.RefKind,
-                isParams: p.IsParams,
-                type: p.Type,
-                name: p.Name,
-                isOptional: true,
-                hasDefaultValue: true));
+                result.Add(new CodeAction.CodeActionWithNestedActions(FeaturesResources.Add_parameter_to_constructor, requiredParameterCodeActions.ToImmutableAndFree(), isInlinable: false));
+                result.Add(new CodeAction.CodeActionWithNestedActions(FeaturesResources.Add_optional_parameter_to_constructor, optionalPrameterCodeActions.ToImmutableAndFree(), isInlinable: false));
+            }
+
+            return result;
+
+            // local functions
+            static CodeAction GetOptionalContructorParametersCodeAction(Document document, ConstructorCandidate constructorCandidate, INamedTypeSymbol containingType, bool useSubMenuName)
+            {
+                var missingOptionalParameters = constructorCandidate._missingParameters.SelectAsArray(p => CodeGenerationSymbolFactory.CreateParameterSymbol(
+                      attributes: default,
+                      refKind: p.RefKind,
+                      isParams: p.IsParams,
+                      type: p.Type,
+                      name: p.Name,
+                      isOptional: true,
+                      hasDefaultValue: true));
+
+                return new AddConstructorParametersCodeAction(document, constructorCandidate, containingType, missingOptionalParameters, useSubMenuName);
+            }
 
-            // return a code action to add optional parameters
-            yield return new AddConstructorParametersCodeAction(this, document, state, missingParameters);
+            static bool CanHaveRequiredParameters(ImmutableArray<IParameterSymbol> parameters)
+                   => parameters.Length == 0 || (parameters.Length > 0 && !parameters.Last().IsOptional);
         }
     }
 }
```

Comment:
skimmed the rest of this file.  will rereview when other changes go in.  thanks!